### PR TITLE
roc analysis

### PR DESCRIPTION
**Pull request by TzipHoro**

### CODE REVIEW

**Comment by elegantmoose** [comment]
@TzipHoro Looking at now

File: ROC.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+        return prec
+
+    def accuracy(self, y_pred: pd.Series) -> float:
+        return accuracy_score(self.y_true, y_pred)
+
+    def f1_score(self, y_pred: pd.Series) -> float:
+        return f1_score(self.y_true, y_pred)
+
+    def threshold_matrix(self, step_size: float) -> pd.DataFrame:
```
Comment (elegantmoose) [inline ROC.py:66]: Nice. For ways to speed up, can def mention making this loop multi-threaded, well multiprocessed to get around python GIL but you know what I  mean.

File: ROC.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+def main(output_data_path: str, plot_path: str = None):
+    df = pd.read_csv(r'data/train.csv', index_col='DataItem')
+    y_test = df.loc[df.index % 2 == 0, 'TACA'].reset_index(drop=True)
+
+    df_output = pd.read_csv(output_data_path, index_col=0)
+    yhat = df_output.loc[:, 'Testing Set 3']
+
+    roc = ROCMetrics(y_test, yhat)
+    thresholds = roc.threshold_matrix(step_size=0.0001)
+    p_th = thresholds.loc[['sensitivity', 'specificity', 'precision', 'accuracy', 'f1_score'], :].sum().idxmax()
```
Comment (elegantmoose) [inline ROC.py:107]: like this

File: ROC.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    return p_th, (yhat >= p_th).astype(int)
+
+
+if __name__ == '__main__':
+    # multi-layer perceptron
+    p_th_1 = main(r'data/multilayer_data/testing_activation_1.csv', r'plots/roc-mlp-1.png')
+    p_th_2 = main(r'data/multilayer_data/testing_activation_2.csv', r'plots/roc-mlp-2.png')
+    p_th_3 = main(r'data/multilayer_data/testing_activation_3.csv', r'plots/roc-mlp-3.png')
+    print('threshold 1:', p_th_1, '\nthreshold 2:', p_th_2, '\nthreshold 3:', p_th_3)
```
Comment (elegantmoose) [inline ROC.py:119]: nitpicking - I wouldnt mind if we also printed out the ROC values here for the chosen optimal threshold

**Review by elegantmoose — APPROVED** [review]
Would it be helpful maybe to see plots with the ROC values for all thresholds? i.e. can see the optimal zones visually. If so I could take a whack at it

**Comment by TzipHoro** [comment]
> Would it be helpful maybe to see plots with the ROC values for all thresholds? i.e. can see the optimal zones visually. If so I could take a whack at it

Since this data set is so small it will probably be hard to visually see the optimal zones even if we use every threshold that was tested, but I'll see what I can do.

**Pull request merged**